src/sctools/metrics/gatherer.py

```diff
@@ -1,41 +1,109 @@
+   GatherCellMetrics            Class to gather metrics on all cells in an experiment
+   GatherGeneMetrics            Class to gather metrics on all genes in an experiment
+
+See Also
+--------
+sctools.metrics.aggregator
+sctools.metrics.merge
+sctools.metrics.writer
+
+"""
+
 from contextlib import closing
 
 import pysam
 
 from sctools.bam import iter_cell_barcodes, iter_genes, iter_molecule_barcodes
-from sctools.metrics.aggregator import CellBarcodeMetrics, GeneMetrics
+from sctools.metrics.aggregator import CellMetrics, GeneMetrics
 from sctools.metrics.writer import MetricCSVWriter
 
 
 class MetricGatherer:
+    """Gathers Metrics from an experiment
 
-    def __init__(self, bam_file: str, output_stem: str, compress: bool=True):
-        """
-        MetricGatherer defines an extract_metrics() function which will parse a file molecule-by-
-        molecule and write the resulting information to an output file.
+    Because molecules tend to have relatively small numbers of reads, the memory footprint of
+    this method is typically small (tens of megabytes).
 
-        Because molecules tend to have relatively small numbers of reads, the memory footprint of
-        this method is typically small (tens of megabytes)
+    Parameters
+    ----------
+    bam_file : str
+        the bam file containing the reads that metrics should be calculated from. Can be a chunk
+        of cells or an entire experiment
+    output_stem : str
+        the file stem for the gzipped csv output
 
-        :param bam_file: the bam file over which to compile metrics
-        :param output_stem: the file stem for the (csv formatted) metric outputs
-        """
+    Methods
+    -------
+    extract_metrics
+        extracts metrics from ``bam_file`` and writes them to output_stem.csv.gz
+
+    """
+
+    def __init__(self, bam_file: str, output_stem: str, compress: bool=True):
         self._bam_file = bam_file
         self._output_stem = output_stem
         self._compress = compress
 
     @property
     def bam_file(self) -> str:
+        """the bam file that metrics are generated from"""
         return self._bam_file
 
-    def extract_metrics(self) -> None:
-        """extract metrics from the provided bam file and write the results to csv."""
+    def extract_metrics(self, mode='rb') -> None:
+        """extract metrics from the provided bam file and write the results to csv.
+
+        Parameters
+        ----------
+        mode : {'r', 'rb'}, default 'rb'
+            the open mode for pysam.AlignmentFile. 'r' indicates the input is a sam file, and 'rb'
+            indicates a bam file.
+
+        """
         raise NotImplementedError
 
 
 class GatherCellMetrics(MetricGatherer):
 
-    # assumes bam file is sorted by GE, UB, CB tags
+    extra_docs = """
+    Notes
+    -----
+    ``bam_file`` must be sorted by gene (``GE``), molecule (``UB``), and cell (``CB``), where gene 
+    varies fastest.
+    
+    Examples
```

Comment:
The examples here look really appealing, though they always cost a lot of time. Again, it can be done gradually in the future.

Reply:
Agree, I think the examples make the code base way more usable. They also let me avoid stupid things. For example, I sometimes feel compelled to write wrappers for existing functions to make usage obvious... examples can solve that problem.